pkg/reconciler/utils/authtype/authtype.go

```diff
@@ -0,0 +1,115 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package authtype
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+
+	"github.com/google/knative-gcp/pkg/reconciler/identity/resources"
+)
+
+const (
+	AuthenticationCheckUnknownReason = "AuthenticationCheckPending"
+	ControlPlaneNamespace            = "cloud-run-events"
+	BrokerServiceAccountName         = "broker"
+)
+
+var BrokerSecret = &corev1.SecretKeySelector{
+	LocalObjectReference: corev1.LocalObjectReference{
+		Name: "google-broker-key",
+	},
+	Key: "key.json",
+}
+
+type AuthTypeArgs struct {
+	Namespace          string
+	ServiceAccountName string
+	Secret             *corev1.SecretKeySelector
+}
+
+func GetAuthType(ctx context.Context, serviceAccountLister corev1listers.ServiceAccountLister,
+	secretLister corev1listers.SecretLister, args AuthTypeArgs) (string, error) {
+	// For AuthTypeArgs from Sources, either ServiceAccountName or Secret will be empty,
+	// because of the IdentitySpec validation from Webhook.
+	// For AuthTypeArgs from BrokerCell, ServiceAccountName and Secret will be both presented.
+	// We need to revisit this function after https://github.com/google/knative-gcp/issues/1888 lands,
+	// which will add IdentitySpec to BrokerCell.
+	// For AuthTypeArgs from BrokerCell.
+	if args.ServiceAccountName != "" && args.Secret != nil {
+		if authType, err := GetAuthTypeForWorkloadIdentity(ctx, serviceAccountLister, args); authType != "" {
+			return authType, err
+		} else if authType, err := GetAuthTypeForSecret(ctx, secretLister, args); authType != "" {
+			return authType, err
+		} else {
+			return "", fmt.Errorf("authentication is not configured, Secret doesn't present, ServiceAccountName doesn't have required annotation")
+		}
+	}
+
+	// For AuthTypeArgs from Sources which has serviceAccountName.
+	if args.ServiceAccountName != "" {
+		return GetAuthTypeForWorkloadIdentity(ctx, serviceAccountLister, args)
+	}
+
+	// For AuthTypeArgs from Sources which has secret.
+	if args.Secret != nil {
+		return GetAuthTypeForSecret(ctx, secretLister, args)
+	}
+
+	return "", fmt.Errorf("invalid AuthTypeArgs, neither ServiceAccountName nor Secret are provided")
+}
+
+func GetAuthTypeForWorkloadIdentity(ctx context.Context, serviceAccountLister corev1listers.ServiceAccountLister, args AuthTypeArgs) (string, error) {
+	kServiceAccount, err := serviceAccountLister.ServiceAccounts(args.Namespace).Get(args.ServiceAccountName)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			return "", fmt.Errorf("using Workload Identity for authentication configuration, " +
+				"can't find Kubernetes Service Account " + args.ServiceAccountName)
+		}
+		return "workload-identity", fmt.Errorf("error getting Kubernests Service Account: %s", err.Error())
```

Comment:
Previously, I use a single GetAuthType function for both Sources and BrokerCell. The scenario "return something with error" is for BrokerCell case. In BrokerCell, I check workload identity first, if I don't need to immediately return error get from it, I will check secret. Some errors from Workload Identity doesn't need to immediately return while some does. (eg. brokercell's service account doesn't have the annotation, it will return error, but it doesn't need to immediately return error, as we will continue to check secret. However, if we have error to get brokerCell's service account from the list, we need to return this error). I used the additional return something to determine wether the error need to be immediately returned or not.

I split GetAuthType function and changed this part's logic. Please check my latest commit.